File: src/mycelos/knowledge/parse_note.py
"""Deterministic DE+EN note parser.

Runs both in the `POST /api/knowledge/notes` handler and (as a sibling
JS port) in the browser Quick-Capture modal. No LLM, no I/O. Given the
same input and reference time, it must return the same result in both
languages and in both runtimes. Shared test vectors in
``tests/fixtures/parse-note-vectors.json`` enforce that.
"""
from __future__ import annotations

import re
from datetime import datetime, timedelta, timezone
from typing import TypedDict


class ParsedNote(TypedDict):
    type: str          # "note" | "task"
    due: str | None    # ISO8601 UTC
    tags: list[str]
    wikilinks: list[str]
    reminder: bool


_WIKILINK_RE = re.compile(r"\[\[([^\]|]+)(?:\|[^\]]+)?\]\]")
_TAG_RE = re.compile(r"(?:^|\s)#([A-Za-zÄÖÜäöüß0-9_-]+)")
_TODO_RE = re.compile(r"^\s*(TODO|FIXME|AUFGABE)\s*[:\s]", re.IGNORECASE)

_REMIND_RE = re.compile(
    r"\b(remind\s+me|erinnere\s+mich|erinner\s+mich)\b",
    re.IGNORECASE,
)
_IN_DURATION_RE = re.compile(
    r"\bin\s+(\d+)\s+(minute|minuten|minutes|min|stunde|stunden|hour|hours|std|h)\b",
    re.IGNORECASE,
)

# "tomorrow 2pm" / "morgen 14 uhr" / "morgen 14:00"
_TOMORROW_RE = re.compile(
    r"\b(tomorrow|morgen)\b(?:\s+(\d{1,2})(?::(\d{2}))?\s*(am|pm|uhr)?)?",
    re.IGNORECASE,
)

# "14.04. 15:00" / "14.04.2026 15:00"
_GERMAN_DATE_RE = re.compile(
    r"\b(\d{1,2})\.(\d{1,2})\.(\d{4})?\s*(\d{1,2})(?::(\d{2}))?",
)


def _to_iso(dt: datetime) -> str:
    return dt.strftime("%Y-%m-%dT%H:%M:%S") + "Z"


def _duration_seconds(amount: int, unit: str) -> int:
    u = unit.lower()
    if u.startswith("min"):
        return amount * 60
    if u.startswith(("h", "std", "stunde", "hour")):
        return amount * 3600
    return amount * 60


def _parse_12h_to_24h(hour: int, ampm: str | None) -> int:
    if not ampm:
        return hour
    ampm = ampm.lower()
    if ampm == "pm" and hour < 12:
        return hour + 12
    if ampm == "am" and hour == 12:
        return 0
    return hour


def parse_note_text(text: str, *, now: datetime | None = None) -> ParsedNote:
    """Parse free-form note text into structured fields. Pure function."""
    if now is None:
        now = datetime.now(tz=timezone.utc)
    if now.tzinfo is None:
        now = now.replace(tzinfo=timezone.utc)

    wikilinks = _WIKILINK_RE.findall(text)
    tags = _TAG_RE.findall(text)

    reminder = bool(_REMIND_RE.search(text))
    type_: str = "note"
    due: str | None = None

    if _TODO_RE.match(text):
        type_ = "task"

    in_dur = _IN_DURATION_RE.search(text)
    if in_dur:
        amount = int(in_dur.group(1))
        seconds = _duration_seconds(amount, in_dur.group(2))
        due = _to_iso(now + timedelta(seconds=seconds))
        type_ = "task"

    if due is None:
        tomorrow = _TOMORROW_RE.search(text)
        if tomorrow:
            hour = int(tomorrow.group(2)) if tomorrow.group(2) else 9
            minute = int(tomorrow.group(3)) if tomorrow.group(3) else 0
            hour = _parse_12h_to_24h(hour, tomorrow.group(4))
            target = (now + timedelta(days=1)).replace(
                hour=hour, minute=minute, second=0, microsecond=0
            )
            due = _to_iso(target)
            type_ = "task"

    if due is None:
        gdate = _GERMAN_DATE_RE.search(text)
        if gdate:
            day = int(gdate.group(1))
            month = int(gdate.group(2))
            year = int(gdate.group(3)) if gdate.group(3) else now.year
            hour = int(gdate.group(4))
            minute = int(gdate.group(5)) if gdate.group(5) else 0
            target = datetime(year, month, day, hour, minute, tzinfo=timezone.utc)
            due = _to_iso(target)
            type_ = "task"

    if reminder and due is None:
        # "remind me" without a time: default to +1h so the caller can still
        # persist the reminder intent. The server may overwrite this.
        due = _to_iso(now + timedelta(hours=1))
        type_ = "task"

    return ParsedNote(
        type=type_,
        due=due,
        tags=tags,
        wikilinks=wikilinks,
        reminder=reminder,
    )

File: src/mycelos/frontend/shared/parse-note.js
// Deterministic DE+EN note parser — client-side twin of
// mycelos/knowledge/parse_note.py. The Python version is the source of
// truth; this module is kept in lockstep via shared test vectors in
// tests/fixtures/parse-note-vectors.json.

const WIKILINK_RE = /\[\[([^\]|]+)(?:\|[^\]]+)?\]\]/g;
const TAG_RE = /(?:^|\s)#([A-Za-zÄÖÜäöüß0-9_-]+)/g;
const TODO_RE = /^\s*(TODO|FIXME|AUFGABE)\s*[:\s]/i;
const REMIND_RE = /\b(remind\s+me|erinnere\s+mich|erinner\s+mich)\b/i;
const IN_DURATION_RE = /\bin\s+(\d+)\s+(minuten|minutes|minute|min|stunden|stunde|hours|hour|std|h)\b/i;
const TOMORROW_RE = /\b(tomorrow|morgen)\b(?:\s+(\d{1,2})(?::(\d{2}))?\s*(am|pm|uhr)?)?/i;
const GERMAN_DATE_RE = /\b(\d{1,2})\.(\d{1,2})\.(\d{4})?\s*(\d{1,2})(?::(\d{2}))?/;

function toIso(date) {
  const pad = (n) => String(n).padStart(2, '0');
  return (
    `${date.getUTCFullYear()}-${pad(date.getUTCMonth() + 1)}-${pad(date.getUTCDate())}` +
    `T${pad(date.getUTCHours())}:${pad(date.getUTCMinutes())}:${pad(date.getUTCSeconds())}Z`
  );
}

function durationSeconds(amount, unit) {
  const u = unit.toLowerCase();
  if (u.startsWith('min')) return amount * 60;
  if (u.startsWith('h') || u.startsWith('std') || u.startsWith('stunde') || u.startsWith('hour')) {
    return amount * 3600;
  }
  return amount * 60;
}

function twelveToTwentyFour(hour, ampm) {
  if (!ampm) return hour;
  const a = ampm.toLowerCase();
  if (a === 'pm' && hour < 12) return hour + 12;
  if (a === 'am' && hour === 12) return 0;
  return hour;
}

export function parseNoteText(text, now = new Date()) {
  const wikilinks = [];
  let m;
  const wRe = new RegExp(WIKILINK_RE.source, 'g');
  while ((m = wRe.exec(text)) !== null) wikilinks.push(m[1]);

  const tags = [];
  const tRe = new RegExp(TAG_RE.source, 'g');
  while ((m = tRe.exec(text)) !== null) tags.push(m[1]);

  const reminder = REMIND_RE.test(text);
  let type = 'note';
  let due = null;

  if (TODO_RE.test(text)) type = 'task';

  const dur = text.match(IN_DURATION_RE);
  if (dur) {
    const amount = parseInt(dur[1], 10);
    const seconds = durationSeconds(amount, dur[2]);
    due = toIso(new Date(now.getTime() + seconds * 1000));
    type = 'task';
  }

  if (due === null) {
    const tm = text.match(TOMORROW_RE);
    if (tm) {
      let hour = tm[2] ? parseInt(tm[2], 10) : 9;
      const minute = tm[3] ? parseInt(tm[3], 10) : 0;
      hour = twelveToTwentyFour(hour, tm[4]);
      const target = new Date(now.getTime());
      target.setUTCDate(target.getUTCDate() + 1);
      target.setUTCHours(hour, minute, 0, 0);
      due = toIso(target);
      type = 'task';
    }
  }

  if (due === null) {
    const gd = text.match(GERMAN_DATE_RE);
    if (gd) {
      const day = parseInt(gd[1], 10);
      const month = parseInt(gd[2], 10) - 1;
      const year = gd[3] ? parseInt(gd[3], 10) : now.getUTCFullYear();
      const hour = parseInt(gd[4], 10);
      const minute = gd[5] ? parseInt(gd[5], 10) : 0;
      due = toIso(new Date(Date.UTC(year, month, day, hour, minute, 0)));
      type = 'task';
    }
  }

  if (reminder && due === null) {
    due = toIso(new Date(now.getTime() + 3600 * 1000));
    type = 'task';
  }

  return { type, due, tags, wikilinks, reminder };
}
